http/src/v0/pin/add.rs

```diff
@@ -0,0 +1,111 @@
+use crate::v0::support::option_parsing::ParseError;
+use crate::v0::support::{StringError, StringSerialized};
+use futures::future::try_join_all;
+use ipfs::{Cid, Ipfs, IpfsTypes};
+use serde::Serialize;
+use std::convert::TryFrom;
+use warp::{reply, Filter, Rejection, Reply};
+
+#[derive(Debug)]
+pub struct AddRequest {
+    args: Vec<Cid>,
+    recursive: bool,
+    progress: bool,
+    // TODO: timeout, probably with rollback semantics?
+}
+
+#[derive(Serialize)]
+struct AddResponse {
+    #[serde(rename = "Pins")]
+    pins: Vec<StringSerialized<Cid>>,
+    // FIXME: go-ipfs doesn't respond with this
+    //progress: u8,
+}
+
+pub async fn add_inner<T: IpfsTypes>(
+    ipfs: Ipfs<T>,
+    request: AddRequest,
+) -> Result<impl Reply, Rejection> {
+    if request.progress {
+        // FIXME: there doesn't appear to be a test for this
+        return Err(crate::v0::support::NotImplemented.into());
+    }
+
+    let cids: Vec<Cid> = request.args;
+
+    let recursive = request.recursive;
+
+    let dispatched_pins = cids.into_iter().map(|x| async {
+        ipfs.insert_pin(&x, recursive)
+            .await
+            .map(move |_| StringSerialized(x))
+    });
+
+    // could be unordered :)
+    let completed = try_join_all(dispatched_pins)
+        .await
+        .map_err(StringError::from)?;
+
+    Ok(reply::json(&AddResponse {
+        pins: completed,
+        //progress: 100,
+    }))
+}
+
+impl<'a> TryFrom<&'a str> for AddRequest {
```

Comment:
nit: same as before - `FromStr` is the idiomatic pick for `TryFrom<&'a str>`; and as before, this is not really a part of the PR, so just putting it out there :stuck_out_tongue: